File: leetcode/lowest_common_ancestor_of_a_binary_search_tree/lowest_common_ancestor_of_a_binary_search_tree.cpp
/*
 * Time Complexity: O(H)
 *   - H is the height of the tree.
 *   - Each recursive call moves down one level of the tree.
 *
 * Space Complexity: O(H)
 *   - Due to the recursion stack.
 *   - In the worst case (skewed tree), H = N.
 */

 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
private:
    int pValue;
    int qValue;

public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == nullptr || root->val == p->val || root->val == q->val) return root;

        pValue = p->val;
        qValue = q->val;

        TreeNode* l = lcaHelper(root->left);
        TreeNode* r = lcaHelper(root->right);

        if(l != nullptr && r != nullptr) return root;
        else if(l != nullptr) return l;
        else return r;
    }

    TreeNode* lcaHelper(TreeNode* root) {
        if(root == nullptr || root->val == pValue || root->val == qValue) return root;

        TreeNode* l = lcaHelper(root->left);
        TreeNode* r = lcaHelper(root->right);

        if(l != nullptr && r != nullptr) return root;
        else if(l != nullptr) return l;
        else return r;
    }
};

File: leetcode/lowest_common_ancestor_of_a_binary_search_tree/lowest_common_ancestor_of_a_binary_search_tree.java
package lowest_common_ancestor_of_a_binary_search_tree;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    private int targetP;
    private int targetQ;

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null || root.val == p.val || root.val == q.val) return root;

        targetP = p.val;
        targetQ = q.val;

        TreeNode l = lcaHelper(root.left);
        TreeNode r = lcaHelper(root.right);

        if(l != null && r != null) return root;
        else if(l != null) return l;
        else return r;
    }

    public TreeNode lcaHelper(TreeNode root) {
        if(root == null || root.val == targetP || root.val == targetQ) return root;

        TreeNode l = lcaHelper(root.left);
        TreeNode r = lcaHelper(root.right);

        if(l != null && r != null) return root;
        else if(l != null) return l;
        else return r;
    }

    public static void main(String[] args) {
        TreeNode tr = new TreeNode(
            6,
            new TreeNode(
                2,
                new TreeNode(0),
                new TreeNode(
                    4,
                    new TreeNode(3),
                    new TreeNode(5)
                )
            ),
            new TreeNode(
                8,
                new TreeNode(7),
                new TreeNode(9)
            )
        );
        new Solution().lowestCommonAncestor(tr, new TreeNode(2), new TreeNode(4));
    }
}
